File: leetcode/18.四数之和.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=18 lang=java
 *
 * [18] 四数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++){
            int newTarget = target - nums[i];
            for(int j=i+1;i<nums.length;j++){
                int l = j+1,r = nums.length-1;
                while (l<r) {
                    int flag = nums[j] + nums[l] + nums[r];
                    if(flag == newTarget){
                        List<Integer> temp = new ArrayList<>();
                        temp.add(nums[i]);
                        temp.add(nums[j]);
                        temp.add(nums[l]);
                        temp.add(nums[r]);
                        if (!res.contains(temp)) {
                                res.add(temp);
                        }
                        l++;
                        r--;
                    }else if(flag > newTarget){
                        r--;
                    }else{
                        l++;
                    }
                }
            }
        }
        return res;
    }

    
}
// @lc code=end

File: 18.四数之和.go
/*
 * @lc app=leetcode.cn id=18 lang=golang
 *
 * [18] 四数之和
 */

// @lc code=start
func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	if len(nums) < 4 {
		return [][]int{}
	}
	res := make([][]int, 0)
	for i := 0; i < len(nums) - 3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for j := i + 1; j < len(nums) - 2; j++ {
			if j > i + 1 && nums[j] == nums[j-1] {
				continue
			}
			l, r := j+1, len(nums)-1
			for l < r {
				if nums[i]+nums[j]+nums[r]+nums[l] == target {
					res = append(res, []int{nums[i], nums[j], nums[l], nums[r]})
					for l < r && nums[l] == nums[l+1] {
						l++
					}
					for l < r && nums[r] == nums[r-1] {
						r--
					}
					l++
					r--
				} else if nums[i]+nums[j]+nums[r]+nums[l] < target {
					l++
				} else {
					r--
				}
			}
		}
	}
	return res
}
// @lc code=end

